lld/ELF/ScriptLexer.cpp

```diff
@@ -185,93 +194,30 @@ StringRef ScriptLexer::skipSpace(StringRef s) {
   }
 }
 
-// An erroneous token is handled as if it were the last token before EOF.
-bool ScriptLexer::atEOF() { return errorCount() || tokens.size() == pos; }
-
-// Split a given string as an expression.
-// This function returns "3", "*" and "5" for "3*5" for example.
-static std::vector<StringRef> tokenizeExpr(StringRef s) {
-  StringRef ops = "!~*/+-<>?^:="; // List of operators
-
-  // Quoted strings are literal strings, so we don't want to split it.
-  if (s.starts_with("\""))
-    return {s};
-
-  // Split S with operators as separators.
-  std::vector<StringRef> ret;
-  while (!s.empty()) {
-    size_t e = s.find_first_of(ops);
-
-    // No need to split if there is no operator.
-    if (e == StringRef::npos) {
-      ret.push_back(s);
-      break;
-    }
-
-    // Get a token before the operator.
-    if (e != 0)
-      ret.push_back(s.substr(0, e));
-
-    // Get the operator as a token.
-    // Keep !=, ==, >=, <=, << and >> operators as a single tokens.
-    if (s.substr(e).starts_with("!=") || s.substr(e).starts_with("==") ||
-        s.substr(e).starts_with(">=") || s.substr(e).starts_with("<=") ||
-        s.substr(e).starts_with("<<") || s.substr(e).starts_with(">>")) {
-      ret.push_back(s.substr(e, 2));
-      s = s.substr(e + 2);
-    } else {
-      ret.push_back(s.substr(e, 1));
-      s = s.substr(e + 1);
-    }
-  }
-  return ret;
-}
-
-// In contexts where expressions are expected, the lexer should apply
-// different tokenization rules than the default one. By default,
-// arithmetic operator characters are regular characters, but in the
-// expression context, they should be independent tokens.
-//
-// For example, "foo*3" should be tokenized to "foo", "*" and "3" only
-// in the expression context.
-//
-// This function may split the current token into multiple tokens.
-void ScriptLexer::maybeSplitExpr() {
-  if (!inExpr || errorCount() || atEOF())
-    return;
-
-  std::vector<StringRef> v = tokenizeExpr(tokens[pos]);
-  if (v.size() == 1)
-    return;
-  tokens.erase(tokens.begin() + pos);
-  tokens.insert(tokens.begin() + pos, v.begin(), v.end());
-}
+// Used to determine whether to stop parsing. Treat errors like EOF.
+bool ScriptLexer::atEOF() { return eof || errorCount(); }
 
 StringRef ScriptLexer::next() {
-  maybeSplitExpr();
-
-  if (errorCount())
-    return "";
-  if (atEOF()) {
-    setError("unexpected EOF");
-    return "";
-  }
-  return tokens[pos++];
+  prevTok = peek();
+  return std::exchange(curTok, StringRef(curBuf.s.data(), 0));
 }
 
 StringRef ScriptLexer::peek() {
```

Comment:
Apologies; didn't get a chance to ask this before it went out.

The lexers I've typically seen preemptively read in the first token; `peek()` is a no-op that returns the previously read token, while `next()` is what reads the next token. Doing it this way creates another bit of state in the lexer: whether or not a token has been read in yet. That creates the issue with atEOF: unless `peek()` has been called at least once, it isn't yet known.

Could the LLD lexer be written that way, or is there a concern that I'm missing? It seems like it would have made less of an upset to the rest of the parser, and it seems simpler to reason about, having less state. (Somewhat subjective, for sure.)

Reply:
The lexer is indeed different from traditional implementations. The reason is to support states (currently just `inExpr` or not).
`peek` caches the token at `curBuf`. `curBuf` might not be used if the state changes. Therefore, `peek` has to check `curBufState` - and prepares to drop the cache if the state changed by the parser.

I think `atEOF` is still like traditional parsers, which require a `peek` to know whether EOF has been encountered.

---

Some `setError` was added in the current way to have minimum impact to the test suite. The `setError` uses could be changed, but many tests would require updates.

Reply:
> The lexer is indeed different from traditional implementations. The reason is to support states (currently just `inExpr` or not). `peek` caches the token at `curBuf`. `curBuf` might not be used if the state changes. Therefore, `peek` has to check `curBufState` - and prepares to drop the cache if the state changed by the parser.
> 
> If `next` is made to read the next token, the token might get stale as well when the state changes.

It seems like this could be addressed by having state changes do a little bit more work: discarding the current token and re-lexing a token in the new state.
 
> I think `atEOF` is still like traditional parsers, which require a `peek` to know whether EOF has been encountered.

This is surprising to me; I had thought that most either maintained a valid atEOF as an invariant, or even called it out as a designated token type. The latter is what I've more typically seen.

Reply:
> It seems like this could be addressed by having state changes do a little bit more work: discarding the current token and re-lexing a token in the new state.

This is exactly what the current `peek()` does.

Are you suggesting that `next()` should consume the current token and peek the next one?
Since we don't always call `peek()` before `next()`, this would require `next()` to call peek twice, which would be awkward.

---

Currently, when `peek()` or `next()` returns the empty string, they indicate an EOF token.

Reply:
> Are you suggesting that `next()` should consume the current token and peek the next one? Since we don't always call `peek()` before `next()`, this would require `next()` to call peek twice, which would be awkward.

Not exactly; I'd expect `peek()` to make no changes to internal state. `next()` would lex a token and set it as the current one, update atEOF, and returning empty string if `atEOF`. It might even be possible to remove `atEOF` by representing this as an empty `curTok`.

I haven't actually tried to put together a patch for this idea yet, so it may well fall apart in the details.

Reply:
I think I get what you meant and I've tried such a solution. Unfortunately, at least with the current uses, `peek() { return curTok; }` isn't feasible. Several places have intricate state transition before `peek`

The most "strange" places:

* readSectionAddressType: `inExpr` is just changed before `peek()`
* `peek() == "=" || peek().starts_with("=")` in `readOutputSectionDescription`
* ...

If we want to make `peek` return curTok, we have to change some `consume` and `expect` to switch the state and there might be some places that are infeasible. I believe the current `expect("("); change state; peek()` way is more readable.